CASSANDRA-17021: Support ZSTD dictionary compression

yifan-c opened the pull request:
Add ZstdDictionaryCompressor and wire to the read and the write path

Resolve a correct dictionary compressor on the read path

lazily init ZstdDictCompress and ZstdDictDecompress

checksumming

Added compression dictionary table and accessors

Wire to System table access to CompressionDictionaryManager

Integrate dictionary trainer

Add nodetool to train dictionary

Cleanup

break CompressionDictionaryManager up

properly release dictionary with reference counting

Add ZstdDictionaryCompressorTest

Simplify trainer interface

Bump SSTable version

Add CompressionDictionaryCacheTest

Add ZstdDictionaryTrainerTest

update to use assertJ

Add CompressionDictionaryTrainingConfigTest

Update comments in CompressionDictionaryEventHandler and update exception handling in MessagingService

Add CompressionDictionaryEventHandlerTest

Add CompressionDictionarySchedulerTest

Minor update

Add CompressionDictionaryManagerTest

Add TrainCompressionDictionaryTest (nodetool)

Add CompressionDictionaryIntegrationTest

Tidy up

Add ZstdCompressionDictionaryTest

Add SystemDistributedKeyspaceCompressionDictionaryTest

Add --sampling-rate for the nodetool

Update consistency level from QUORUM to ONE for best effort

Add the missing setCompressionDictionaryManager to sstable writer builders

Add benchmark

yifan-c committed:
Support ZSTD dictionary compression

Add ZstdDictionaryCompressor and wire to the read and the write path

Resolve a correct dictionary compressor on the read path

lazily init ZstdDictCompress and ZstdDictDecompress

checksumming

Added compression dictionary table and accessors

Wire to System table access to CompressionDictionaryManager

Integrate dictionary trainer

Add nodetool to train dictionary

Cleanup

break CompressionDictionaryManager up

properly release dictionary with reference counting

Add ZstdDictionaryCompressorTest

Simplify trainer interface

Bump SSTable version

Add CompressionDictionaryCacheTest

Add ZstdDictionaryTrainerTest

update to use assertJ

Add CompressionDictionaryTrainingConfigTest

Update comments in CompressionDictionaryEventHandler and update exception handling in MessagingService

Add CompressionDictionaryEventHandlerTest

Add CompressionDictionarySchedulerTest

Minor update

Add CompressionDictionaryManagerTest

Add TrainCompressionDictionaryTest (nodetool)

Add CompressionDictionaryIntegrationTest

Tidy up

Add ZstdCompressionDictionaryTest

Add SystemDistributedKeyspaceCompressionDictionaryTest

Add --sampling-rate for the nodetool

Update consistency level from QUORUM to ONE for best effort

Add the missing setCompressionDictionaryManager to sstable writer builders

Add benchmark

# Conflicts:
#	src/java/org/apache/cassandra/net/Verb.java

yifan-c committed:
Add configs to cassandra.yaml and refactoring

yifan-c committed:
Handle CFS invalidate event and close the CompressionDictionaryManager

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ZstdDictionaryTrainer.java:
```diff
@@ -0,0 +1,334 @@
+                        currentSampleCount, totalSampleSize.get(), isReady);
+            byte[] dictBytes = zstdTrainer.trainSamples();
+            long zstdDictId = Zstd.getDictIdFromDict(dictBytes);
+            DictId dictId = new DictId(Kind.ZSTD, makeDictionaryId(System.currentTimeMillis(), zstdDictId));
```
Avoid System for time, should use `org.apache.cassandra.utils.Clock.Globa`l or `org.apache.cassandra.utils.Clock` interface [blockSystemClock]

smiklosovic commented on src/java/org/apache/cassandra/tools/nodetool/TrainCompressionDictionary.java:
```diff
@@ -0,0 +1,188 @@
+                        "writing to new SSTable, you might consider running nodetool 'flush' along " +
+                        "with this command to have chunk available for sampling)");
+            long maxWaitMillis = TimeUnit.SECONDS.toMillis(maxSamplingDurationSeconds + 300); // Add 5 minutes for training
+            long startTime = System.currentTimeMillis();
```
Avoid System for time, should use `org.apache.cassandra.utils.Clock.Global` or `org.apache.cassandra.utils.Clock` interface [blockSystemClock]

smiklosovic commented on src/java/org/apache/cassandra/tools/nodetool/TrainCompressionDictionary.java:
```diff
@@ -0,0 +1,188 @@
+            long maxWaitMillis = TimeUnit.SECONDS.toMillis(maxSamplingDurationSeconds + 300); // Add 5 minutes for training
+            long startTime = System.currentTimeMillis();
+            
+            while (System.currentTimeMillis() - startTime < maxWaitMillis)
```
same

smiklosovic commented on conf/cassandra.yaml:
```diff
@@ -617,6 +617,54 @@ counter_cache_save_period: 7200s
+compression_dictionary_refresh_interval: 3600s
+
+# Initial delay before starting the first dictionary refresh cycle after node startup.
+# This prevents all nodes from refreshing simultaneously when the cluster starts.
```
That means that all nodes would need to be started at the very same moment? Is this planned to be different per node in a real scenario? Because if all nodes have 10s then ... all of them would have same delay set, hence we would have the "simultaneous" situation again? 

Maybe having _random initial delay_ in some _range_ (e.g. 5 to 20 seconds) would be better?

yifan-c replied:
I do not think starting all nodes at the same moment is a requirement. The refresh interval just defines how frequent each node should check the system table. When there is a new dictionary trained, the node will broadcast the message to all live nodes to notify the dictionary is available. This happens outside of the periodic polling. 
Combining both, you can consider the interval as the upper bound of how dictionary is refreshed (when broadcasting fails at some nodes). 
When some nodes are running with the prior dictionary or no dictionary, it is still not a huge issue, from the perspective of read and write SSTables. Eventually, all nodes will pick up the latest.

rustyrazorblade replied:
What is the overhead of this?  This seems like a huge bikeshed over something that has extremely low overhead.

yifan-c replied:
@rustyrazorblade, I am not sure if I understand your comment. I would like to keep the settings as they are for now and not introduce randomized delay unless we see an actual need. If the default interval turns out to be sub-optimal, we can adjust that directly.

rustyrazorblade replied:
Sorry for the confusion.  I'm not advocating for randomized delay.  The point I was trying to make is that this operation appears to be so lightweight that debating if it should be run once a day or an hour seems like a bikeshed.  We're talking about an incredibly trivial amount of data and a very lightweight operation.  Doing it once a day vs an hour or even a minute is like the difference between one and three grain of sand on an entire beach.

smiklosovic commented on conf/cassandra.yaml:
```diff
@@ -617,6 +617,54 @@ counter_cache_save_period: 7200s
+# How long to keep compression dictionaries in the cache before they expire.
+# Expired dictionaries will be removed from memory but can be reloaded if needed.
+# Min unit: s
+compression_dictionary_cache_expire: 3600s
```
probably safe just saying the fact that this is same (3600s) as `compression_dictionary_refresh_interval` feels like there might be some kind of a problem when both periods ends at the same time.

rustyrazorblade replied:
Why such an aggressive expiration time?  I'd make this 24h.

yifan-c replied:
`compression_dictionary_cache_expire` and `compression_dictionary_refresh_interval` are not strongly connected, since they apply to different components. Cache entry expiry is renewed on every access. Meanwhile, `compression_dictionary_refresh_interval` only defines how frequent to read from system table.
If there is no access to the cache entry for 1 hour, it is evicted. The chance is low that the dictionary is wanted right after eviction. Even in such scenario, the cache entry is again populated from compressionInfo or system table.

yifan-c replied:
I am open to 24 hours. Was just placing some default values in the initial patch.

rustyrazorblade replied:
I think the eviction code is broken, see my full comment in JIRA.  When compacting a table that I left overnight (assuming the dictionary was evicted) I get this:

```
WARN  [CompactionExecutor:869] 2025-10-08T17:16:05,033 SSTableWriter.java:157 - Failed to open /Users/jhaddad/dev/cassandra/bin/../data/data/chunk4_2/keyvalue-1b255f4def2540a600000000000000f6/pa-2-big for writing
java.lang.IllegalStateException: Dictionary is being closed
    at org.apache.cassandra.io.compress.ZstdDictionaryCompressor.lambda$getOrCreate$1(ZstdDictionaryCompressor.java:100)
    ```

yifan-c replied:
This should be fixed in https://github.com/apache/cassandra/pull/4399/commits/1ae168a2efc86aba50a74a3de4fb180ab3916112.

smiklosovic commented on src/java/org/apache/cassandra/db/compression/CompressionDictionaryCache.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    @Override
+    public synchronized void close()
```
not sure if it is possible but what happens if somebody adds to this cache while this method is being executed. I think the fact this is synchronized just means that two threads can not execute `close` method at the same time but it does not mean that other thread could not call other methods in this cache. 

Would not it be better to do the logic around closing `currentDictionary` in `removalListener` directly? Then we would just call `cache.invalidateAll();` and were done with it. 

Also, what happens when we go to remove a dictionary by the means of `removalListener` while that dictionary is set in `currentDictionary`?

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ManualTrainingOptions.java:
```diff
@@ -0,0 +1,80 @@
+ * Configuration options for manual compression dictionary training.
+ * This class encapsulates the parsed and validated parameters needed for training.
+ */
+public class ManualTrainingOptions
```
nit: braces not necessary at some places

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ManualTrainingOptions.java:
```diff
@@ -0,0 +1,80 @@
+    {
+        if (options == null || !options.containsKey("maxSamplingDurationSeconds"))
+        {
+            throw new IllegalArgumentException("maxSamplingDurationSeconds parameter is required for manual dictionary training");
```
Can not that be some default when missing? Seems like a hassle to be force to include it all the time.

yifan-c replied:
The nodetool command is supposed to provide a value, always. So the exception should never be thrown. I had been weighing both options for a while; the reason of not providing a default here is that 1) it is not going to be used, and 2) two default values in nodetool and server could diverge and create confusion.

rustyrazorblade replied:
What about using a max # of samples instead of a duration?  Duration is kind of weird, highly variable.  Then a default also makes sense.

yifan-c replied:
The reason of using duration is to have a time bound for sampling. If using the max # of samples, there is a chance that the tool runs too long, even indefinitely, if no flush and no compaction occurs.
Having a time bound at least will terminate the tool process in a controlled manner. I acknowledge that having a failed training after 10 minutes is frustrating. It is up to operators to configure the tool properly; also the SSTable-based sampling that you suggested and I just added should solve the issue better.

smiklosovic replied:
@yifan-c well with a big gun there comes a lot of responsibility :) I would also introduce number of samples, if we are worried about that never ending, I would add duration _as additional boundary_. 

For example, max samples 100 000 OR duration of 6 hours. Whatever happens first will end it. 

@rustyrazorblade  what do you think?

rustyrazorblade replied:
I think the priority should be reducing the number of operator headaches.  If it runs for 6 hours, would most people want it to keep running or fail?  I'd rather it keep going.  What is the benefit of failing, ever?  What is the problem if it takes a full week?  

>  I acknowledge that having a failed training after 10 minutes is frustrating. It is up to operators to configure the tool properly;

I think we have a responsibility not to ship foot-guns.  Defaults should work well.  Expecting people to change every config under the guise of "operator responsibility" is a mindset that caused Cassandra to ship some pretty awful defaults for years - GC being a notable example.  It's done a lot of damage to the project, and I see it as an abdication of an important responsibility.

yifan-c replied:
The typical reason for training failure (after about 10 minutes) is that no `flush` is invoked with the `traincompressiondictionary` command. Otherwise, within 10 minutes, there should normally be enough samples collected.
To clarify the current implementation: the termination condition for training is already defined as either reaching the maximum number of samples or the maximum sampling duration. The maximum sample size (`compression_dictionary_training_max_total_sample_size`) is configured in `cassandra.yaml`. The trainer starts the dictionary training once either of these thresholds is reached.
I agree that we should avoid shipping foot-guns. Based on the feedback, I see problems with the write-based sampling approach. It is too complex since it requires running another command (`flush`) to ensure success. Frequent flushes can also lead to undesirable side effects, such as producing many small SSTables.
On the other hand, SSTable-based sampling works better for manual training. It provides a more deterministic and self-contained process, and ensures the command terminates without having to worry about the sampling duration or manual intervention.
In contrast, write-based sampling seems more suitable for automatic (auto-training) scenarios.
Given these points, should we remove write-based sampling from the manual `traincompressiondictionary` command? What do you think?

smiklosovic replied:
SSTable-based sampling - that one which also solves Jon's usecase by training existing data? Yeah I think that is more robust. My instinct is that by write-based training we are trying to be too smart, while it works more "real-time-ish" and more suitable for automatic scenarios, it has also unpredictable behaviour if not enough writes or similar when used manually.

rustyrazorblade replied:
I think if starting training first triggers a flush, meaning we'll almost certainly have data on disk (unless there are zero writes ever, which we can't do much about), then removing training from the write path seems perfectly reasonable to me.  I can't think of a reason to keep the write based sampling if we flush first.

yifan-c replied:
Sounds good. I am open to remove the write-based sampling from the manual training tool invocation. 

I would like to clarify on whether we want to run `flush` on the table regardless or only when there is no live sstable. I think the former is probably fine as the training tool is used very sparsely, possibly only once.

smiklosovic replied:
that's a good question. What do you do when there is 10 SSTables already (or if somebody is changing compression algorithm)? You take the last one (newest one) or the first one? (oldest one?) I would take the newest one. 

What if you flush and that flush will not have enough data? Like it would flush too few data. Then you take the older one?

yifan-c replied:
Stefan, please check out https://github.com/apache/cassandra/commit/f7a66b47f382ade1fb689135056e2690f128a136 first. The sstable sampling is already implemented in this commit. It basically does random sampling across all data on disk. What it does not have is running 'flush' in advance.

rustyrazorblade replied:
I think if there's no SSTables on disk, run flush.  Otherwise, train on what's there.  

@yifan-c thanks for explaining the sampling, that's what I had in mind too.

smiklosovic replied:
@yifan-c thanks, I had  to overlook that.

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ManualTrainingOptions.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    public static ManualTrainingOptions fromStringMap(Map<String, String> options)
+    {
+        if (options == null || !options.containsKey("maxSamplingDurationSeconds"))
```
are we really going to use  camel case? not snake case? not sure what is better or what we use consistently. Also we might probably extract this string to a standalone constant in this class. Easy to mess it up by typing it constantly.

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ZstdCompressionDictionary.java:
```diff
@@ -0,0 +1,217 @@
+    public ZstdDictCompress dictionaryForCompression(int compressionLevel)
+    {
+        if (closed.get())
+            throw new IllegalStateException("Dictionary has been closed");
```
maybe being more descriptive in this message (and all other same messages) by including dict id with kind and id would be nice.

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ZstdCompressionDictionary.java:
```diff
@@ -0,0 +1,217 @@
+        @Override
+        public String name()
+        {
+            return "ZstdCompressionDictionary";
```
`ZstdCompressionDictionary.class.getSimpleName()`.

smiklosovic commented on src/java/org/apache/cassandra/io/compress/ZstdDictionaryCompressor.java:
```diff
@@ -0,0 +1,216 @@
+            })
+            .build();
+
+    // dictioanry and its ref are null, when they are absent.
```
typo in `dictioanry`.

smiklosovic commented on src/java/org/apache/cassandra/io/compress/IDictionaryCompressor.java:
```diff
@@ -0,0 +1,71 @@
+ * 
+ * @param <T> the specific type of compression dictionary this compressor supports
+ */
+public interface IDictionaryCompressor<T extends CompressionDictionary>
```
https://github.com/apache/cassandra/pull/4399/files#r2410732332

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ManualTrainingOptions.java:
```diff
@@ -0,0 +1,80 @@
+
+    public ManualTrainingOptions(int maxSamplingDurationSeconds)
+    {
+        if (maxSamplingDurationSeconds <= 0)
```
consider `Preconditions.checkArgument` here

smiklosovic commented on src/java/org/apache/cassandra/io/compress/CompressionMetadata.java:
```diff
@@ -139,11 +165,54 @@ private CompressionMetadata(CompressionMetadata copy)
+        // 1. The current compressor is not a dictionary compressor, but there is dictionary attached
+        // 2. The current dictionary compressor is a different type, e.g. table schema is changed
+        // In those cases, we should get the compatible dictionary compressor based on the dictionary
+        if (dictionary.kind() == ZSTD)
```
We are again too "ZSTD-centric" here. I would expect this does not contain anything ZSTD specific so we do not need to touch this anymore. The fact that dictionary is not empty again pretty much guarantees that we can create a compressor from it, no? Why would we have a dictionary, of `CompressionDictionary` type, if we were not able to create a compressor from it? That would be a pretty useless dictionary to me. 

So maybe doing something like 

`return dictionary.kind().getCompressor(dictionary);`

and calling `ZstdDictionaryCompressor.create(zstdDict)` in there would be way more simpler and at one place. 

You would need to return `ICompressor` from `getCompressor` as `resolveCompressor` returns it too. If `IDictionaryCompressor` extended `ICompressor` as already suggested elsewhere (as you need to instantiate `ZstdDictionaryCompressor` there) it would be a piece of cake.

smiklosovic commented on src/java/org/apache/cassandra/db/compression/CompressionDictionaryManager.java:
```diff
@@ -0,0 +1,346 @@
+    public synchronized void train(Map<String, String> options)
+    {
+        // Validate table supports dictionary compression
+        if (!isEnabled)
```
Preconditions.checkArgument ?

smiklosovic commented on src/java/org/apache/cassandra/db/compression/CompressionDictionaryManager.java:
```diff
@@ -0,0 +1,346 @@
+        ManualTrainingOptions trainingOptions = ManualTrainingOptions.fromStringMap(options);
+
+        trainer.start(true);
+        scheduler.scheduleManualTraining(trainingOptions, trainer);
```
we would not actually _scheduled_ anything in that method if trainer is no-op instance.

smiklosovic commented on src/java/org/apache/cassandra/db/compression/CompressionDictionaryManager.java:
```diff
@@ -0,0 +1,346 @@
+    public synchronized void close()
+    {
+        unregisterMbean();
+        if (trainer != null)
```
closing no-op trainer would do nothing

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ICompressionDictionaryTrainer.java:
```diff
@@ -0,0 +1,168 @@
+        }
+
+        IDictionaryCompressor dictionaryCompressor = (IDictionaryCompressor) compressor;
+        if (dictionaryCompressor.acceptableDictionaryKind() == CompressionDictionary.Kind.ZSTD)
```
we are zstd specific here, not sure that is really needed.

smiklosovic commented on src/java/org/apache/cassandra/tools/NodeProbe.java:
```diff
@@ -2682,6 +2683,61 @@ public void setMixedMajorVersionRepairEnabled(boolean enabled)
+    private CompressionDictionaryManagerMBean getDictionaryManagerProxy(String keyspace, String table) throws IOException
+    {
+        // Construct table-specific MBean name
+        String mbeanName = "org.apache.cassandra.db.compression:type=CompressionDictionaryManager,keyspace=" + keyspace + ",table=" + table;
```
replace with constant once introduced

rustyrazorblade committed:
Added docs for zstd dict

smiklosovic commented on src/java/org/apache/cassandra/db/compression/CompressionDictionary.java:
```diff
@@ -0,0 +1,210 @@
+     *
+     * @return dictionary id
+     */
+    DictId identifier();
```
I would rename to dictId();

smiklosovic committed:
review

yifan-c committed:
Yifan's refactoring on top of Stefan's

yifan-c committed:
Revert the long to DictId change from Stefan

yifan-c committed:
Revert "Revert the long to DictId change from Stefan"

This reverts commit db95727cb70fd8bfee805a82f143d44d5eb54452.

yifan-c committed:
Address Stefan's review comments

Get rid of java.util.concurrent.CompletableFuture usage

Use clock

Relocate dictionary configurations and use IntKibibytesBound

Change compression_dictionary_training_sampling_rate data type from int to float

Address other various comments

yifan-c commented:
I think all comments are addressed. Please tag me if I missed. (GH is not super friendly with many comments)

yifan-c committed:
Fix currentDictionary being closed prematurely

yifan-c committed:
Display statistics when running traincompressiondictionary

smiklosovic committed:
resolve remaining checkstyle issues

smiklosovic commented on src/java/org/apache/cassandra/db/compression/ZstdDictionaryTrainer.java:
```diff
@@ -0,0 +1,341 @@
+            currentTrainingStatus = TrainingStatus.FAILED;
+            String errorMsg = String.format("Insufficient samples for training: %d (minimum required: %d)",
+                                            currentSampleCount, MIN_SAMPLES_REQUIRED);
+            throw new IllegalStateException(errorMsg);
```
This will be visible in server's logs:

java.lang.IllegalStateException: Insufficient samples for training: 0 (minimum required: 10)
	at org.apache.cassandra.db.compression.ZstdDictionaryTrainer.trainDictionary(ZstdDictionaryTrainer.java:119)
	at org.apache.cassandra.db.compression.ICompressionDictionaryTrainer.lambda$trainDictionaryAsync$0(ICompressionDictionaryTrainer.java:80)
	at org.apache.cassandra.concurrent.FutureTask.call(FutureTask.java:61)
	at org.apache.cassandra.concurrent.FutureTask.run(FutureTask.java:71)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)


I am not sure we want to have this visible like that, we should log it only.

smiklosovic replied:
I would change the result of this whole method to "TraningResult" where status would be only one component of that.

smiklosovic committed:
rework nodetool command to use table builder

smiklosovic committed:
reuse constant

smiklosovic committed:
log just message instead of whole stacktrace

yifan-c committed:
Add --use-existing-sstables option in traincompressiondictionary

yifan-c committed:
minor change

yifan-c committed:
Define DEFAULT_SAMPLING_DURATION_SECONDS

yifan-c committed:
simplify the close method in CompressionDictionaryCache

yifan-c committed:
Fix issues in SSTableChunkSampler

yifan-c committed:
Remove write-based sampling from nodetool traincompressiondictionary

yifan-c commented:
Pushed changes and trigger the CI https://pre-ci.cassandra.apache.org/job/cassandra/123/pipeline-overview/

yifan-c committed:
Update traincompressiondictionary command

- remove async mode and status options, as the sstable-based sampling is deterministic and fast
- propagate the failure message to command output, if the runs fails
- update compression.adoc

yifan-c commented:
A new commit to further simplify/enhance the nodetool command. And a new CI run https://pre-ci.cassandra.apache.org/job/cassandra/124/

yifan-c committed:
Print error when table not found

yifan-c committed:
Fix mbean not found issue

yifan-c committed:
Add ZstdDictionaryCompressor for auto-completion

yifan-c committed:
Only register mbean when dictionary compression is enabled

yifan-c committed:
Fix tests

yifan-c commented on src/java/org/apache/cassandra/utils/StorageCompatibilityMode.java:
```diff
@@ -35,6 +35,13 @@ public enum StorageCompatibilityMode
      */
     CASSANDRA_4(4),
 
+    /**
+     * Similar to CASSANDRA_4.
+     * The new features in 6.0 are
+     * - ZSTD dictionary-based compression. Once SSTables are compressed with dictionary, they cannot be rolled back.
+     */
+    CASSANDRA_5(5),
+
```
Is it necessary? `StorageCompatibilityMode` is to check the sstable format compatibility, which is not changed in this patch. Also, the dictionary feature is opt-in. I'd incline removing the code. It seems unnecessary. 

cc: @smiklosovic , @rustyrazorblade

yifan-c committed:
Add --force option to traincompressiondictionary

yifan-c committed:
Avoid triggering JVMStabilityInspector error logs

yifan-c committed:
Make a command group

yifan-c committed:
Increase time precision from minutes to seconds in the dictionary ID by using a custom time anchor

yifan-c committed:
fix tests

yifan-c committed:
Persist dictionary checksum and length

yifan-c committed:
update changes.txt

yifan-c commented:
Thanks all. The patch is merged!
